File: solution/Solution2.py
import bisect
from typing import List


class Solution:
    # Alternative solution for performance testing
    def numTeams(self, rating: List[int]) -> int:
        # Initialize an empty list to store the ratings in sorted order
        l = []
        
        # Create a sorted copy of the rating array
        sr = sorted(rating)
        
        # Create a dictionary to store the indices of the sorted ratings
        low = {}
        for idx, r in enumerate(sr):
            low[r] = idx
        
        # Initialize the result variable
        res = 0
        
        # Get the length of the rating array
        n = len(rating)
        
        # Iterate through the rating array
        for idx, r in enumerate(rating):
            # Find the insertion point of the current rating in the sorted list
            i = bisect.bisect(l, r)
            
            # Insert the current rating at the correct position in the sorted list
            l.insert(i, r)
            
            # Calculate the number of teams that can be formed with the current rating
            j = low[r] - i
            res += i * (n - 1 - idx - j) + j * (idx - i)
        
        # Return the total number of teams
        return res

    # Second solution
    def numTeams2(self, rating: List[int]) -> int:
        # Initialize the result variable
        ans = 0
        
        # Get the length of the rating array
        n = len(rating)
        
        # Iterate through the rating array
        for j in range(n):
            # Initialize variables to count the number of ratings less than and greater than the current rating
            llt, lgt = 0, 0
            
            # Count the number of ratings less than and greater than the current rating in the range [0..j)
            for i in range(j):
                llt += rating[i] < rating[j]
                lgt += rating[i] > rating[j]
            
            # Initialize variables to count the number of ratings less than and greater than the current rating
            rlt, rgt = 0, 0
            
            # Count the number of ratings less than and greater than the current rating in the range (j..n)
            for k in range(j + 1, n):
                rlt += rating[k] < rating[j]
                rgt += rating[k] > rating[j]
            
            # Calculate the number of teams that can be formed with the current rating
            ans += llt * rgt + lgt * rlt
        
        # Return the total number of teams
        return ans

    # Brute force solution (time limit exceeded)
    def numTeams1(self, rating: List[int]) -> int:
        # Initialize the result variable
        ans = 0
        
        # Get the length of the rating array
        n = len(rating)
        
        # Iterate through all possible combinations of three soldiers
        for i in range(n):
            for j in range(i + 1, n):
                for k in range(j + 1, n):
                    # Check if the rating conditions are met
                    ans += 1 if rating[i] < rating[j] < rating[k] or rating[i] > rating[j] > rating[k] else 0
        
        # Return the total number of teams
        return ans

File: solution/Solution2.ts
// numTeams.ts
/**
 * Function to calculate the number of teams that can be formed from a given rating array.
 * A team consists of three elements in increasing or decreasing order.
 * 
 * @param rating The input rating array
 * @returns The total count of teams that can be formed
 */
function numTeams(rating: number[]): number {
    // Initialize a variable to store the total count of teams
    let count = 0;
    
    // Get the length of the rating array
    const n = rating.length;
    
    // Iterate through the rating array
    for (let j = 0; j < n; j++) {
        // Initialize variables to store the count of less and more ratings on the left and right sides
        let leftLess = 0, leftMore = 0, rightLess = 0, rightMore = 0;
        
        // Iterate through the elements before the current element (left side)
        for (let i = 0; i < j; i++) {
            // If the current element is greater than the previous element, increment the leftLess count
            // This means the previous element is less than the current element
            if (rating[i] < rating[j]) leftLess++;
            // If the current element is less than the previous element, increment the leftMore count
            // This means the previous element is more than the current element
            if (rating[i] > rating[j]) leftMore++;
        }
        
        // Iterate through the elements after the current element (right side)
        for (let k = j + 1; k < n; k++) {
            // If the current element is greater than the next element, increment the rightLess count
            // This means the next element is less than the current element
            if (rating[k] < rating[j]) rightLess++;
            // If the current element is less than the next element, increment the rightMore count
            // This means the next element is more than the current element
            if (rating[k] > rating[j]) rightMore++;
        }
        
        // Calculate the count of teams that can be formed with the current element as the middle element
        // This is done by multiplying the count of less ratings on the left with the count of more ratings on the right
        // and adding the count of more ratings on the left with the count of less ratings on the right
        // This represents the number of teams that can be formed with the current element as the middle element
        count += leftLess * rightMore + leftMore * rightLess;
    }
    
    // Return the total count of teams
    return count;
}